### PR TITLE
Rianhughes/07 get block with receipts

### LINKED ISSUE #516: RPC07 implement new method

### PR DESCRIPTION
**Pull request by rianhughes**
closes https://github.com/NethermindEth/starknet.go/issues/516

Edit: This PR incorporates the changes from rc-1 which relate to block receipts

### CODE REVIEW

File: rpc/types_block.go [hunk 1]
```diff
@@ -123,6 +123,28 @@ type PendingBlock struct {
...
+	BlockBodyWithReceipts
+}
+
+type BlockBodyWithReceipts struct {
+	Transactions []TransactionWithReceipt `json:"transactions"`
+}
+
+type TransactionWithReceipt struct {
+	Transaction Transaction        `json:"transaction,omitempty"`
+	Receipt     TransactionReceipt `json:"receipt"`
```
Comment (omerfirmak) [inline rpc/types_block.go]: This field's schema is not yet finalized, let's wait that to be finalized.

Comment (rianhughes) [reply rpc/types_block.go:139]: Receipts were updated in rc1. I will incorporate the rc-1 changes in this PR

Comment (rianhughes) [reply rpc/types_block.go]: Receipts seem to be finalised in rc-1. I'll incorporate them into this PR.

**Comment by rianhughes** [comment]
Todo: update tests when juno supports rpc v07

**Comment by IronGauntlets** [comment]
The PR, in general, looks good, however, it still needs to be updated as per your comment. I can have a look again after you have updated it.

File: rpc/block.go [hunk 2]
```diff
@@ -166,3 +167,25 @@ func (provider *Provider) BlockWithTxs(ctx context.Context, blockID BlockID) (in
...
+	if err := do(ctx, provider.c, "starknet_getBlockWithReceipts", &result, blockID); err != nil {
+		return nil, tryUnwrapToRPCErr(err, ErrBlockNotFound)
+	}
+	var block BlockWithReceipts
+	err := json.Unmarshal(result, &block)
+	if err == nil && block.BlockStatus != "" {
+		return &block, nil
+	}
+
+	// if result.Status == nil it's a pending block
```
Comment (IronGauntlets) [inline rpc/block.go]: This is incorrect, please see [here](https://github.com/starkware-libs/starknet-specs/blob/7ae9e7bdcd2126bd30830b96b10932cf59d13ae3/api/starknet_api_openrpc.json#L3201). If the block is pending then the status should be `"PENDING"`.

File: rpc/types_block_test.go [hunk 3]
```diff
@@ -125,3 +127,48 @@ func TestBlock_Unmarshal(t *testing.T) {
...
+			blockCasted := block.(*BlockWithReceipts)
+			txsWithReceipts := blockCasted.BlockBodyWithReceipts.Transactions
+			require.Equal(t, blockCasted.BlockStatus, test.ExpectedBlock.Status)
+			require.Equal(t, blockCasted.BlockNumber, test.ExpectedBlock.BlockNumber)
+			require.NotZero(t, len(blockCasted.Transactions))
+			require.NotNil(t, blockCasted.L1GasPrice)
+			require.NotNil(t, blockCasted.L1DataGasPrice)
+			require.NotNil(t, txsWithReceipts)
+			require.NotNil(t, txsWithReceipts[0].Transaction)
+			require.NotNil(t, txsWithReceipts[0].Receipt)
```
Comment (IronGauntlets) [inline rpc/types_block_test.go]: If `test` and `blockCasted` have the same fields and type then you can do `require.Equal`

File: rpc/types_transaction_receipt.go [hunk 4]
```diff
@@ -475,6 +475,19 @@ func (tr *UnknownTransactionReceipt) UnmarshalJSON(data []byte) error {
...
+		recInner, err := json.Marshal(dec["TransactionReceipt"])
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal(recInner, &decInner); err != nil {
+			return err
+		}
+		dec = decInner
+	}
+
```
Comment (IronGauntlets) [inline rpc/types_transaction_receipt.go]: It seems to me like the only difference between this piece of code and the one lines 257-268 in file `rpc/types_transaction.go` is `Transaction` and `TransactionReceipt`. Please consider removing this duplication.

File: rpc/block.go (hunk 2 above)
Comment (rianhughes) [reply rpc/block.go]: getBlockWithReceipts can return a [BLOCK_WITH_RECEIPTS](https://github.com/starkware-libs/starknet-specs/blob/7ae9e7bdcd2126bd30830b96b10932cf59d13ae3/api/starknet_api_openrpc.json#L120)which has a [BLOCK_STATUS](https://github.com/starkware-libs/starknet-specs/blob/7ae9e7bdcd2126bd30830b96b10932cf59d13ae3/api/starknet_api_openrpc.json#L1634) (which may be "PENDING"), however getBlockWithReceipts can also [return](https://github.com/starkware-libs/starknet-specs/blob/7ae9e7bdcd2126bd30830b96b10932cf59d13ae3/api/starknet_api_openrpc.json#L1679) a distinct type [PENDING_BLOCK_WITH_RECEIPTS](https://github.com/starkware-libs/starknet-specs/blob/7ae9e7bdcd2126bd30830b96b10932cf59d13ae3/api/starknet_api_openrpc.json#L1679) which has no BLOCK_STATUS.

Comment (rianhughes) [reply rpc/block.go]: The naming could be clearer tbh (on the spec side), but I think the current implementation is correct.

Comment (rianhughes) [reply rpc/block.go]: I've updated the logic to make this a bit clearer, but it's essentially the same

File: rpc/types_block_test.go (hunk 3 above)
Comment (rianhughes) [reply rpc/types_block_test.go]: Updated

File: rpc/types_transaction_receipt.go (hunk 4 above)
Comment (rianhughes) [reply rpc/types_transaction_receipt.go]: I've moved this duplicated logic to a helper

File: rpc/block.go (hunk 2 above)
Comment (IronGauntlets) [reply rpc/block.go]: Yeah looks like the spec is ambiguous

**Review by IronGauntlets — APPROVED** [review]

**Pull request merged**